### ISSUE TITLE
False Positive for needless_collect with two mutable borrows

### ISSUE THREAD

**Issue by HoshigaIkaro** [rust-lang/rust-clippy#10035]
### Summary

The applied lint fails to compile due to a `Map` created with items created from mutable borrows.

### Lint Name

clippy::needless_collect

### Reproducer

I tried this code:

```rust
...
let v: Vec<char> = (0..num)
    .map(|_| cmap.get_mut(origin - 1).unwrap().pop().unwrap())
    .collect();
cmap.get_mut(target - 1)
    .unwrap()
    .extend(v.into_iter().rev());
...
}
```

I saw this happen:

```
warning: avoid using `collect()` when not needed
  --> src\days\day_05.rs:28:18
   |
28 |                 .collect();
   |                  ^^^^^^^
...
31 |                 .extend(v.into_iter().rev());
   |                         ------------- the iterator could be used here instead
   |
   = note: `#[warn(clippy::needless_collect)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect
help: use the original Iterator instead of collecting it and then producing a new one
   |
26 ~
27 |             cmap.get_mut(target - 1)
28 |                 .unwrap()
29 ~                 .extend((0..num)
30 ~                 .map(|_| cmap.get_mut(origin - 1).unwrap().pop().unwrap()).rev());
   |
```

I expected to see this happen:

No warning. 

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-pc-windows-msvc
release: 1.65.0
LLVM version: 15.0.0
```

**Comment by J-ZhengLi** [comment]
Rather than completely silence this lint for this scenario, anyone have any suggestion how to write this in a optimal way?

**Comment by mickvangelderen** [comment]
> Rather than completely silence this lint in this scenario, anyone have any suggestion how to write this in an optimal way?

@J-ZhengLi I guess this is off-topic but I didn't want to leave you hanging.

The original example is incomplete but it seems like the implemented operation tries to move some number of elements under one index to another index. I imagine it does something like this:

```rust
fn do_the_thing(cmap: &mut Vec<Vec<char>>, origin: usize, target: usize, num: usize) {
    let origin_vec = cmap.get_mut(origin).unwrap();
    let moved_chars = origin_vec.split_off(origin_vec.len().saturating_sub(num));
    cmap.get_mut(target).unwrap().extend(moved_chars);
}
```

The reason for allocating a temporary vector is that the map type (`Vec<_>` in my example) does not make it easy to get multiple mutable references to different elements at simultaneously. It can be done though by leveraging the iterator implementation of `Vec`. 

```rust
fn swap_if<T>(swap: bool, values: (T, T)) -> (T, T) {
    if swap {
        (values.1, values.0)
    } else {
        values
    }
}

fn vec_index_mut_2<T>(vec: &mut Vec<T>, indices: (usize, usize)) -> (Option<&mut T>, Option<&mut T>) {
    if indices.0 == indices.1 {
        return (vec.get_mut(indices.0), None)
    }

    let (low_index, high_index) = swap_if(indices.0 > indices.1, indices);
    
    let mut iter = vec.iter_mut().skip(low_index);
    let low_item = iter.next();
    let mut iter = iter.skip(high_index - (low_index + 1));
    let high_item = iter.next();

    swap_if(indices.0 > indices.1, (low_item, high_item))
}

fn do_the_thing2(cmap: &mut Vec<Vec<char>>, origin: usize, target: usize, num: usize) {
    if let (Some(origin_vec), Some(target_vec)) = vec_index_mut_2(cmap, (origin, target)) {
        target_vec.extend(origin_vec.drain(origin_vec.len().saturating_sub(num)..));
    }
}
```

It may also be implemented by splitting a mutable slice. I have no idea what the performance implications are. Alternatively you can try to find extensions to the standard library, other data types, or use run-time borrow checking with `Cell` or `RefCell`.
